### PR TITLE
builder pattern draft

### PR DESCRIPTION
**Pull request by louis-bompart**

### CODE REVIEW

File: samples/pushBatchOfDocumentsFromFolder.ts [hunk 1]
```diff
@@ -33,12 +33,10 @@ async function main() {
     '/path/to/fileB.json',
   ];
 
-  const {onBatchUpload, onBatchError, done} =
-    await source.batchUpdateDocumentsFromFiles('my_source_id', entries);
-
-  onBatchUpload(onSuccessCallback);
-  onBatchError(onErrorCallback);
-  await done();
+  await source
```
Comment (y-lakhdar) [inline samples/pushBatchOfDocumentsFromFolder.ts:36]: You are missing an await.
`batchUpdateDocumentsFromFiles` returns a promise, so you cannot really use the builder pattern here.
image

Comment (y-lakhdar) [inline samples/pushBatchOfDocumentsFromFolder.ts:36]: You are missing an await.
`batchUpdateDocumentsFromFiles` returns a promise, so you cannot really use the builder pattern here.
image

File: src/source/batchUpdateDocumentsFromFile.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        for (const filePath of files.values()) {
+          const docBuilders =
+            parseAndGetDocumentBuilderFromJSONDocument(filePath);
+          await analyser.add(docBuilders);
+        }
+        await pushSource.createFields(analyser);
+      }
+      await this.batchConsumer.consume(files);
+    }).bind(this)();
+    this.then(() => this.internalPromise);
```
Comment (y-lakhdar) [inline src/source/batchUpdateDocumentsFromFile.ts:51]: This causes a `RangeError: Maximum call stack size exceeded` error

**Comment by y-lakhdar** [comment]
Did manual tests and it works as expected.
Let's merge it

**Review by y-lakhdar — APPROVED** [review]

File: src/source/batchUpdateDocumentsFromFile.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import {FileConsumer} from '../help/fileConsumer';
+import {
+  BatchUpdateDocuments,
+  BatchUpdateDocumentsFromFiles,
+  FailedUploadCallback,
+  FieldAnalyser,
+  SuccessfulUploadCallback,
+} from '../index';
+import {parseAndGetDocumentBuilderFromJSONDocument} from '../validation/parseFile';
+import {PushSource} from './push';
```
Comment (y-lakhdar) [inline src/source/batchUpdateDocumentsFromFile.ts:11]: let's just remove the circular dependency

Comment (louis-bompart) [reply src/source/batchUpdateDocumentsFromFile.ts:11]: ? You need it to ref' PushSource from this file.

If you want to remove it you'll need to rewire/refactor a lot more code.
You can replace it by an `import type` tho

**Pull request merged**